Drop WebKit's fallback memmem() implementation and move sub-span search to WTF::find()

iidmsa opened the pull request:
#### c2659c292ffc8da008e82d1847770cbfbd219e19
<pre>
Drop WebKit&apos;s fallback memmem() implementation and move sub-span search to WTF::find()
<a href="https://bugs.webkit.org/show_bug.cgi?id=301151">https://bugs.webkit.org/show_bug.cgi?id=301151</a>
<a href="https://rdar.apple.com/163075589">rdar://163075589</a>

Reviewed by Yusuke Suzuki.

Remove custom memmem() implementation that handled platforms without native memmem()
support (e.g., Windows) by implementing a span-based search algorithm. On platforms with
native memmem(), WTF::find() still delegates to the optimized system-provided implementation.
This only affects Windows and other platforms lacking memmem().

* Source/WTF/wtf/StdLibExtras.h:
(WTF::find):
(WTF::contains):
(WTF::memmem): Deleted.
* Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:
(WebCore::CurlMultipartHandle::findBoundary):
(WebCore::CurlMultipartHandle::parseHeadersIfPossible):

Canonical link: <a href="https://commits.webkit.org/301949@main">https://commits.webkit.org/301949@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0608abd53be033b380c479c1df27974d00ac4228

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127518 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47166 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38305 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134606 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79074 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129390 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47784 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55691 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97077 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65025 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130466 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38218 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114204 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77557 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/37045 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32303 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77968 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/119555 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108081 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32734 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137079 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/125981 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54177 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41747 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105595 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54688 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110562 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105246 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26842 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50761 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29208 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51757 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54114 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60201 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/159019 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53348 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39748 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56805 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55107 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -1088,18 +1062,37 @@ size_t find(std::span<T, TExtent> haystack, std::span<U, UExtent> needle)
     static_assert(sizeof(T) == sizeof(U));
+
+#if !HAVE(MEMMEM)
+    if (!needle.size())
```
empty() would be easy to read

Constellation commented on Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:
```diff
@@ -203,26 +204,28 @@ CurlMultipartHandle::FindBoundaryResult CurlMultipartHandle::findBoundary()
 
-    auto boundaryLength = m_boundary.length();
-    auto boundaryStartPtr = m_boundary.data();
+    auto boundarySpan = std::span { reinterpret_cast<const uint8_t*>(m_boundary.data()), m_boundary.length() };
```
Are there any utility function for this pattern? (If not, then the current one is fine)

cdumez replied:
I think we'd normally use `m_boundary.span()` with a `byteCast()`to convert from `const char` to `const uint8_t`.

cdumez commented on Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:
```diff
@@ -203,26 +204,28 @@ CurlMultipartHandle::FindBoundaryResult CurlMultipartHandle::findBoundary()
         if (!m_didCompleteMessage) {
             // Not enough data to check the boundary (Temporarily retain "Initial CRLF + (boundary - 1)" bytes for the next search.)
-            result.dataEnd = std::max(int(contentLength) - int(boundaryLength) + 1 - 2, 0);
+            result.dataEnd = std::max(int(contentLength) - int(boundarySpan.size()) + 1 - 2, 0);
```
We should use `static_cast<int>()` instead of `int()`

iidmsa replied:
Done.

iidmsa committed:
Drop WebKit's fallback memmem() implementation and move sub-span search to WTF::find()
https://bugs.webkit.org/show_bug.cgi?id=301151
rdar://163075589

Reviewed by Yusuke Suzuki.

Remove custom memmem() implementation that handled platforms without native memmem()
support (e.g., Windows) by implementing a span-based search algorithm. On platforms with
native memmem(), WTF::find() still delegates to the optimized system-provided implementation.
This only affects Windows and other platforms lacking memmem().

* Source/WTF/wtf/StdLibExtras.h:
(WTF::find):
(WTF::contains):
(WTF::memmem): Deleted.
* Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:
(WebCore::CurlMultipartHandle::findBoundary):
(WebCore::CurlMultipartHandle::parseHeadersIfPossible):

Canonical link: https://commits.webkit.org/301949@main
